prov/efa/src/efa_av.c

```diff
@@ -1075,3 +1075,102 @@ int efa_av_open(struct fid_domain *domain_fid, struct fi_av_attr *attr,
 	free(av);
 	return ret;
 }
+
+#ifdef EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+
+/* Common Setup/Teardown */
+
+struct efa_resource {
+	struct fid_fabric *fabric;
+	struct fid_domain *domain;
+	struct fid_av *av;
+};
+
+struct efa_resource *create_efa_resource() {
+    struct efa_resource *resource = (struct efa_resource*) malloc(sizeof(struct efa_resource));
+    struct fi_info *info;
+    struct fi_av_attr av_attr = {0};
+    int ret;
+
+    resource->fabric = NULL;
+    resource->domain = NULL;
+    resource->av = NULL;
+
+    ret = fi_getinfo(FI_VERSION(1, 14), NULL, NULL, 0ULL, NULL, &info);
+    if (ret) {
+        free(resource);
+        return NULL;
+    }
+
+    ret = fi_fabric(info->fabric_attr, &resource->fabric, NULL);
+    if (ret) {
+        free(resource);
+        return NULL;
+    }
+
+    ret = fi_domain(resource->fabric, info, &resource->domain, NULL);
+    if (ret) {
+        fi_close((fid_t) resource->fabric);
+        free(resource);
+        return NULL;
+    }
+
+    ret = fi_av_open(resource->domain, &av_attr, &resource->av, NULL);
+    if (ret) {
+        fi_close((fid_t) resource->domain);
+        fi_close((fid_t) resource->fabric);
+        free(resource);
+        return NULL;
+    }
+
+    return resource;
+}
+
+void delete_efa_resource(struct efa_resource *resource) {
+    fi_close((fid_t) resource->av);
+    fi_close((fid_t) resource->domain);
+    fi_close((fid_t) resource->fabric);
+    free(resource);
+}
+
+/*
+ * Only works on nodes with EFA devices
+ * This test calls efa_ah_alloc twice with the same GID,
+ * and verifies that ibv_create_ah only gets called once.
+ */
+void create_efa_ah_twice() {
+    /* Setup Local Variables */
+    struct efa_resource *resource = create_efa_resource();
+    struct efa_av *av;
+    int ibv_ret = 4242;
+    struct efa_ah *efa_ah;
+    struct efa_ah *efa_ah2;
+
+    assert_non_null(resource);
+    av = container_of(resource->av, struct efa_av, util_av.av_fid);
+    assert_non_null(av);
+    uint8_t gid[EFA_GID_LEN];
```

Comment:
definition of `gid` should be at the beginning of the function.

Reply:
done